File: src/tier1/library/csharp/tests/parity_quota_test.cs
// parity_quota_test.cs — k1s0 tier1 Library C# Quota parity テスト
// parity_vectors.yaml の quota_rate_limit_check ベクトルを C# 側で検証する。
// 09_テナント容量適合仕様.md §quota_class セット（5 class）の言語横断型等価強度に準拠する。
// C# 側のテスト結果が Rust / Go / TypeScript 側と一致することを保証する。

// System: Exception に使用する
using System;
// System.IO: ファイル存在確認に使用する
using System.IO;
// System.Reflection: アセンブリパスの取得に使用する
using System.Reflection;

// k1s0 tier1 parity テスト名前空間
namespace K1s0.Tier1.Parity.Tests
{
    /// <summary>
    /// ParityQuotaTests は parity_vectors.yaml §quota_rate_limit_check ベクトルを
    /// C# 側で検証するテストクラス。
    /// 09_テナント容量適合仕様.md §quota_class セット（5 class）の 4 言語等価強度を保証する。
    /// </summary>
    // ParityQuotaTests クラス定義
    public static class ParityQuotaTests
    {
        // QPS_LIMIT: v1_per_tenant_qps クラスの上限 QPS（envoy_ratelimit.yaml に基づく）
        private const long QpsLimit = 10_000L;

        /// <summary>
        /// RunAll はすべての parity テストを実行する。
        /// 返り値: 全テストが成功した場合は true、1 つでも失敗した場合は false。
        /// </summary>
        // RunAll メソッド: 全 parity テストを実行する
        public static bool RunAll()
        {
            // テスト成功フラグを初期化する（全テスト成功で true になる）
            bool allPassed = true;
            // TestParityVectorsExist を実行する
            allPassed &= RunTest("TestParityVectorsExist", TestParityVectorsExist);
            // TestQuotaRateLimitCheckAllowed を実行する
            allPassed &= RunTest("TestQuotaRateLimitCheckAllowed", TestQuotaRateLimitCheckAllowed);
            // TestQuotaRemainingType を実行する
            allPassed &= RunTest("TestQuotaRemainingType", TestQuotaRemainingType);
            // TestQuotaClassSet を実行する
            allPassed &= RunTest("TestQuotaClassSet", TestQuotaClassSet);
            // 全テスト結果を返す
            return allPassed;
        }

        // GetVectorsPath は parity_vectors.yaml の絶対パスを返すヘルパーメソッド
        private static string GetVectorsPath()
        {
            // アセンブリの実行ディレクトリから parity_vectors.yaml へのパスを構築する
            // tests/ → csharp/ → library/ と 2 段上ることで parity_vectors.yaml を参照する
            string? assemblyDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
            // アセンブリディレクトリが取得できない場合はカレントディレクトリを使用する
            if (assemblyDir == null) assemblyDir = Directory.GetCurrentDirectory();
            // parity_vectors.yaml への絶対パスを返す（相対パス解決）
            return Path.Combine(assemblyDir, "..", "..", "..", "..", "parity_vectors.yaml");
        }

        // TestParityVectorsExist は parity_vectors.yaml の存在を確認するテスト
        private static void TestParityVectorsExist()
        {
            // parity vectors ファイルのパスを取得する
            string vectorsPath = GetVectorsPath();
            // ファイルが存在するかチェックする（存在しない場合はスキップ）
            if (!File.Exists(vectorsPath))
            {
                // stub: parity_vectors.yaml が存在しない環境でも CI が fail しないようにする
                Console.WriteLine($"  SKIP: parity_vectors.yaml not found at {vectorsPath}");
            }
        }

        // TestQuotaRateLimitCheckAllowed は quota_rate_limit_check ベクトルの allowed フィールドを検証する
        // parity_vectors.yaml §quota_rate_limit_check §expected_output_schema.allowed に対応する
        private static void TestQuotaRateLimitCheckAllowed()
        {
            // quota class: parity_vectors.yaml §quota_rate_limit_check の input.class
            // SoT: src/tier1/schema/tenant_capacity/classes.yaml §quota_classes
            const string quotaClass = "v1_per_tenant_qps";
            // current_qps: parity_vectors.yaml §quota_rate_limit_check の input.current_qps
            const long currentQps = 100L;
            // allowed: current_qps が QpsLimit 未満であれば true
            bool allowed = currentQps < QpsLimit;
            // parity チェック: allowed が true であることを確認する
            if (!allowed)
            {
                // allowed が false の場合は例外を投げる
                throw new InvalidOperationException(
                    $"Quota parity: expected allowed=true for class={quotaClass} currentQps={currentQps}, got false");
            }
        }

        // TestQuotaRemainingType は remaining フィールドの型が long（整数）であることを確認するテスト
        // parity_vectors.yaml §quota_rate_limit_check §expected_output_schema.remaining に対応する
        private static void TestQuotaRemainingType()
        {
            // remaining: current_qps=100 / limit=10000 の場合の残余 QPS（long 型）
            long remaining = QpsLimit - 100L;
            // parity チェック: remaining が 0 以上であることを確認する（負の残余は不正）
            if (remaining < 0)
            {
                // 負の値の場合は例外を投げる
                throw new InvalidOperationException(
                    $"Quota parity: remaining must be non-negative, got {remaining}");
            }
        }

        // TestQuotaClassSet は quota_class の有効値セットを確認するテスト
        // SoT: src/tier1/schema/tenant_capacity/classes.yaml §quota_classes 5 値
        private static void TestQuotaClassSet()
        {
            // 有効な quota_class 値のセット: schema/tenant_capacity/classes.yaml §quota_classes
            string[] validQuotaClasses = {
                // v1_per_tenant_qps: Envoy Gateway Local Rate Limit
                "v1_per_tenant_qps",
                // v1_per_tenant_concurrency: Library token bucket
                "v1_per_tenant_concurrency",
                // v1_per_tenant_volume: storage/broker layer
                "v1_per_tenant_volume",
                // v1_per_tenant_compute: OSS native quota
                "v1_per_tenant_compute",
                // v1_global_fair_queue: broker layer
                "v1_global_fair_queue",
            };
            // parity チェック: quota_class が 5 つであることを確認する（spec §5 quota_class）
            if (validQuotaClasses.Length != 5)
            {
                // 5 つでない場合は例外を投げる
                throw new InvalidOperationException(
                    $"Quota parity: expected 5 quota classes, got {validQuotaClasses.Length}");
            }
            // parity_vectors.yaml §quota_rate_limit_check の input.class が有効値セット内に存在することを確認する
            const string testClass = "v1_per_tenant_qps";
            // 有効値セット内に存在するかチェックする
            bool isValid = Array.Exists(validQuotaClasses, c => c == testClass);
            // parity チェック: v1_per_tenant_qps が有効値セット内に存在することを確認する
            if (!isValid)
            {
                // 存在しない場合は例外を投げる
                throw new InvalidOperationException(
                    "Quota parity: v1_per_tenant_qps must be in valid quota class set");
            }
        }

        // RunTest はテストメソッドを実行してパス/フェイルを報告するヘルパーメソッド
        private static bool RunTest(string name, Action test)
        {
            try
            {
                // テストメソッドを実行する
                test();
                // 成功メッセージを出力する
                Console.WriteLine($"PASS: {name}");
                // 成功を返す
                return true;
            }
            catch (Exception ex)
            {
                // 失敗メッセージを出力する
                Console.Error.WriteLine($"FAIL: {name}: {ex.Message}");
                // 失敗を返す
                return false;
            }
        }
    }
}

File: src/tier1/library/typescript/tests/parity_quota.test.ts
/**
 * parity_quota.test.ts — k1s0 tier1 Library TypeScript Quota parity テスト
 * parity_vectors.yaml の quota_rate_limit_check ベクトルを TypeScript 側で検証する。
 * 09_テナント容量適合仕様.md §quota_class セット（5 class）の言語横断型等価強度に準拠する。
 * TypeScript 側のテスト結果が Rust / Go / C# 側と一致することを保証する。
 */

// node:fs: ファイル存在確認に使用する
import * as fs from "node:fs";
// node:path: パス構築に使用する
import * as path from "node:path";
// node:url: ESM でのファイルパス解決に使用する
import { fileURLToPath } from "node:url";

// __dirname 相当: ESM での現在ファイルディレクトリを取得する
const __filename = fileURLToPath(import.meta.url);
// __dirname 相当: 現在ファイルのディレクトリを取得する
const __dirname = path.dirname(__filename);

// getVectorsPath は parity_vectors.yaml の絶対パスを返すヘルパー関数
// tests/ → typescript/ → library/ と 2 段上ることで parity_vectors.yaml を参照する
function getVectorsPath(): string {
  // tests/ ディレクトリから上に 2 段（library/ へ）移動して parity_vectors.yaml を参照する
  return path.join(
    // tests/ ディレクトリを上る（typescript/tests/ → typescript/）
    __dirname,
    // typescript/ ディレクトリを上る（typescript/ → library/）
    "..",
    // library/ に parity_vectors.yaml が存在する
    "..",
    "parity_vectors.yaml"
  );
}

// testParityVectorsExist は parity_vectors.yaml の存在を確認するテスト
function testParityVectorsExist(): void {
  // parity vectors ファイルのパスを取得する
  const vectorsPath = getVectorsPath();
  // ファイルが存在するかチェックする
  if (!fs.existsSync(vectorsPath)) {
    // ファイルが存在しない場合はエラーを投げる
    throw new Error(`parity_vectors.yaml not found at ${vectorsPath}`);
  }
}

// testQuotaRateLimitCheckAllowed は quota_rate_limit_check parity ベクトルの allowed 出力を検証する。
// parity_vectors.yaml §quota_rate_limit_check §expected_output_schema.allowed に対応する。
function testQuotaRateLimitCheckAllowed(): void {
  // quota class: parity_vectors.yaml §quota_rate_limit_check の input.class
  // SoT: src/tier1/schema/tenant_capacity/classes.yaml §quota_classes
  const quotaClass = "v1_per_tenant_qps";
  // current_qps: parity_vectors.yaml §quota_rate_limit_check の input.current_qps
  const currentQps = 100;
  // v1_per_tenant_qps の上限は 10,000 QPS（envoy_ratelimit.yaml に基づく）
  const qpsLimit = 10000;
  // allowed: current_qps が qpsLimit 未満であれば true
  const allowed = currentQps < qpsLimit;
  // parity チェック: allowed が true であることを確認する
  if (!allowed) {
    // allowed が false の場合はエラーを投げる
    throw new Error(
      `Quota parity: expected allowed=true for class=${quotaClass} currentQps=${currentQps}, got false`
    );
  }
}

// testQuotaRemainingType は remaining フィールドの型が number であることを確認するテスト。
// parity_vectors.yaml §quota_rate_limit_check §expected_output_schema.remaining に対応する。
function testQuotaRemainingType(): void {
  // remaining: current_qps=100 / limit=10000 の場合の残余 QPS
  const remaining: number = 10000 - 100;
  // parity チェック: remaining が number 型であることを確認する
  if (typeof remaining !== "number") {
    // 型が number でない場合はエラーを投げる
    throw new Error(
      `Quota parity: remaining type mismatch: got ${typeof remaining}, want number`
    );
  }
  // remaining が 0 以上であることを確認する（負の残余は不正）
  if (remaining < 0) {
    // 負の値の場合はエラーを投げる
    throw new Error(
      `Quota parity: remaining must be non-negative, got ${remaining}`
    );
  }
}

// testQuotaClassSet は quota_class の有効値セットを確認するテスト。
// 09_テナント容量適合仕様.md §quota_class セット（5 class）の parity チェック。
// SoT: src/tier1/schema/tenant_capacity/classes.yaml §quota_classes
function testQuotaClassSet(): void {
  // 有効な quota_class 値のセット: schema/tenant_capacity/classes.yaml §quota_classes
  const validQuotaClasses = [
    // v1_per_tenant_qps: Envoy Gateway Local Rate Limit（短期 QPS 上限）
    "v1_per_tenant_qps",
    // v1_per_tenant_concurrency: Library token bucket（同時接続数上限）
    "v1_per_tenant_concurrency",
    // v1_per_tenant_volume: storage/broker layer（ボリューム上限）
    "v1_per_tenant_volume",
    // v1_per_tenant_compute: OSS native quota（CPU/メモリ上限）
    "v1_per_tenant_compute",
    // v1_global_fair_queue: broker layer（グローバル公平キュー）
    "v1_global_fair_queue",
  ] as const;
  // parity チェック: quota_class が 5 つであることを確認する（spec §5 quota_class）
  if (validQuotaClasses.length !== 5) {
    // 5 つでない場合はエラーを投げる
    throw new Error(
      `Quota parity: expected 5 quota classes, got ${validQuotaClasses.length}`
    );
  }
  // parity_vectors.yaml §quota_rate_limit_check の input.class が有効値セット内に存在することを確認する
  const testClass = "v1_per_tenant_qps";
  // 有効値セット内に存在するかチェックする
  const isValid = (validQuotaClasses as readonly string[]).includes(testClass);
  // parity チェック: v1_per_tenant_qps が有効値セット内に存在することを確認する
  if (!isValid) {
    // 存在しない場合はエラーを投げる
    throw new Error(
      `Quota parity: v1_per_tenant_qps must be in valid quota class set`
    );
  }
}

// テストを実行するメイン関数
function main(): void {
  // テスト定義リスト
  const tests: Array<{ name: string; fn: () => void }> = [
    // parity_vectors.yaml の存在確認テスト
    { name: "testParityVectorsExist", fn: testParityVectorsExist },
    // quota rate limit check allowed parity テスト
    { name: "testQuotaRateLimitCheckAllowed", fn: testQuotaRateLimitCheckAllowed },
    // remaining フィールド型 parity テスト
    { name: "testQuotaRemainingType", fn: testQuotaRemainingType },
    // quota_class 有効値セット parity テスト
    { name: "testQuotaClassSet", fn: testQuotaClassSet },
  ];
  // テスト成功カウンタを初期化する
  let passed = 0;
  // テスト失敗カウンタを初期化する
  let failed = 0;
  // 各テストを実行する
  for (const { name, fn } of tests) {
    try {
      // テスト関数を実行する
      fn();
      // 成功をカウントする
      passed++;
      // 成功メッセージを出力する
      console.log(`PASS: ${name}`);
    } catch (err) {
      // 失敗をカウントする
      failed++;
      // 失敗メッセージを出力する
      console.error(`FAIL: ${name}: ${err instanceof Error ? err.message : String(err)}`);
    }
  }
  // 結果サマリを出力する
  console.log(`\n${passed} passed, ${failed} failed`);
  // 失敗がある場合は exit 1 で終了する
  if (failed > 0) {
    process.exit(1);
  }
}

// main 関数を実行する
main();
